File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        // 트리의 노드 중 해당하는 노드의 val값이 맞으면 그 노드를 지운다.
        if (root == null) {
            return null;
        }

        if (root.val > key) {
            root.left = deleteNode(root.left, key);
        } else if (root.val < key) {
            root.right = deleteNode(root.right, key);
        } else {
            // 노드의 val가 키랑 값이 같을때
            // 1.자식이 없을 때(리프노드)
            if (root.left == null && root.right == null) {
                // 그 노드를 지운다.
                return null;
            }
            // 2. 자식이 하나만 있을 때
            if (root.left == null) {
                return root.right;
            } else if (root.right == null) {
                return root.left;
            }
            // 3. 자식이 둘다있을 때
            // 그러면 지운노드가 맞을 때 그 노드의 자식 노드 중 오른쪽의 것들 중 가장 작은노드를
            if (root.left != null && root.right != null) {
                // root의 right 중 가장 작은 노드를 찾는다.
                TreeNode result = findNode(root.right);
                // 지운 노드로 끌어올려 노드의 지워진 부분을 메꾼다.
                root.val = result.val;
                // root 오른쪽 하위노드를 지운다.
                root.right = deleteNode(root.right, result.val);
            }
        }
        return root;
    }

    private TreeNode findNode(TreeNode node) {
        while (node.left != null) {
            node = node.left;
        }        
        return node;
    }
}

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root:
            return None
        # root.val > key -> 왼쪽
        if root.val > key:
            root.left = self.deleteNode(root.left, key)
        # root.val < key -> 오른쪽
        elif root.val < key:
            root.right = self.deleteNode(root.right, key)
        # 같으면 자식들 탐방
        else:
            # 리프노드
            if not root.right and not root.left:
                return None
            
            # 자식하나
            if not root.left:
                return root.right
            elif not root.right:
                return root.left
            
            # 자식이 둘
            def findNode(node: TreeNode) -> TreeNode:
                while node.left:
                    node = node.left
                return node
            # 오른쪽 자식중에 제일 작은 애를 root의 val로 만든다.
            result = findNode(root.right)
            root.val = result.val
            root.right = self.deleteNode(root.right, result.val)

        return root
        